main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala

```diff
@@ -0,0 +1,852 @@
+  }
+
+  private def compileException(visitor: MethodVisitor, className: JvmName, msg: String): Unit = {
+    visitor.visitTypeInsn(NEW, className.toInternalName)
+    visitor.visitInsn(DUP)
+    visitor.visitLdcInsn(msg)
+    // TODO: Load actual source location or change the exception
+    visitor.visitFieldInsn(GETSTATIC, "ca/uwaterloo/flix/language/ast/package$SourceLocation$", "MODULE$", "Lca/uwaterloo/flix/language/ast/package$SourceLocation$;")
+    visitor.visitMethodInsn(INVOKEVIRTUAL, "ca/uwaterloo/flix/language/ast/package$SourceLocation$", "Unknown", "()Lca/uwaterloo/flix/language/ast/package$SourceLocation;", false)
+    visitor.visitMethodInsn(INVOKESPECIAL, className.toInternalName, "<init>", "(Ljava/lang/String;Lca/uwaterloo/flix/language/ast/package$SourceLocation;)V", false)
+    visitor.visitInsn(ATHROW)
+  }
+
+  /*
+   * Generate code to load an integer constant.
+   *
+   * Uses the smallest number of bytes necessary, e.g. ICONST_0 takes 1 byte to load a 0, but BIPUSH 7 takes 2 bytes to
+   * load a 7, and SIPUSH 200 takes 3 bytes to load a 200. However, note that values on the stack normally take up 4
+   * bytes. The exception is if we set `isLong` to true, in which case a cast will be performed if necessary.
+   *
+   * This is needed because sometimes we expect the operands to be a long, which means two (int) values are popped from
+   * the stack and concatenated to form a long.
+   */
+  private def compileInt(visitor: MethodVisitor)(i: Long, isLong: Boolean = false): Unit = {
+    i match {
+      case -1 => visitor.visitInsn(ICONST_M1)
+      case 0 => if (!isLong) visitor.visitInsn(ICONST_0) else visitor.visitInsn(LCONST_0)
+      case 1 => if (!isLong) visitor.visitInsn(ICONST_1) else visitor.visitInsn(LCONST_1)
+      case 2 => visitor.visitInsn(ICONST_2)
+      case 3 => visitor.visitInsn(ICONST_3)
+      case 4 => visitor.visitInsn(ICONST_4)
+      case 5 => visitor.visitInsn(ICONST_5)
+      case _ if scala.Byte.MinValue <= i && i <= scala.Byte.MaxValue => visitor.visitIntInsn(BIPUSH, i.toInt)
+      case _ if scala.Short.MinValue <= i && i <= scala.Short.MaxValue => visitor.visitIntInsn(SIPUSH, i.toInt)
+      case _ if scala.Int.MinValue <= i && i <= scala.Int.MaxValue => visitor.visitLdcInsn(i.toInt)
+      case _ => visitor.visitLdcInsn(i)
+    }
+    if (isLong && scala.Int.MinValue <= i && i <= scala.Int.MaxValue && i != 0 && i != 1) visitor.visitInsn(I2L)
+  }
+
+  private def compileUnaryExpr(currentClassType: JvmType.Reference,
+                               visitor: MethodVisitor,
+                               jumpLabels: Map[Symbol.LabelSym, Label],
+                               entryPoint: Label)(op: UnaryOperator, e: Expression)(implicit root: Root, flix: Flix): Unit = {
+    // Adding source line number for debugging
+    addSourceLine(visitor, e.loc)
+
+    compileExpression(currentClassType, e, jumpLabels, entryPoint, visitor)
+    op match {
```

Comment:
Use the semantic operator instead of the unary operator.

This also gets rid of issues with type.

Reply:
I think we should use use both.